File: r_scripts/forms/rra_291_gross_premium_ibnr.R
# RRA 291 Gross Premium and IBNR Form - R Script
# This script processes RRA 291 Gross Premium and IBNR estimates
# R port of python_scripts/forms/rra_291_gross_premium_ibnr.py

library(dplyr)
library(tidyr)
library(readr)


#' Process RRA 291 Gross Premium and IBNR data
#'
#' @param data_source Path to the data CSV file
#' @return data.frame with processed data ready for analysis
#' @export
process_rra_291 <- function(data_source = '../../synthetic_data/rra_291_gross_premium_ibnr.csv') {

  # Load data
  df <- readr::read_csv(data_source, show_col_types = FALSE)

  # Calculate total incurred
  df <- df %>%
    mutate(
      Total_Incurred_Gross = Paid_Claims_Gross + Case_Reserves_Gross + IBNR_Best_Estimate
    )

  # Calculate reserve ratios
  df <- df %>%
    mutate(
      Case_Reserve_Ratio = ifelse(
        Total_Incurred_Gross > 0,
        Case_Reserves_Gross / Total_Incurred_Gross,
        0
      ),
      IBNR_Ratio = ifelse(
        Total_Incurred_Gross > 0,
        IBNR_Best_Estimate / Total_Incurred_Gross,
        0
      ),
      Paid_Ratio = ifelse(
        Total_Incurred_Gross > 0,
        Paid_Claims_Gross / Total_Incurred_Gross,
        0
      )
    )

  # Calculate IBNR range
  df <- df %>%
    mutate(
      IBNR_Range = IBNR_High - IBNR_Low,
      IBNR_Range_Pct = ifelse(
        IBNR_Best_Estimate > 0,
        IBNR_Range / IBNR_Best_Estimate,
        0
      )
    )

  # Calculate variance from best estimate
  df <- df %>%
    mutate(
      IBNR_High_Variance = IBNR_High - IBNR_Best_Estimate,
      IBNR_Low_Variance = IBNR_Best_Estimate - IBNR_Low
    )

  # Calculate ultimate loss and loss ratio
  df <- df %>%
    mutate(
      Ultimate_Loss = Total_Incurred_Gross,
      Earned_Premium_Ratio = ifelse(
        Gross_Written_Premium > 0,
        Gross_Earned_Premium / Gross_Written_Premium,
        0
      )
    )

  # Calculate loss ratio metrics
  df <- df %>%
    mutate(
      Loss_Ratio_Incurred = ifelse(
        Gross_Earned_Premium > 0,
        Total_Incurred_Gross / Gross_Earned_Premium,
        0
      ),
      Loss_Ratio_Paid = ifelse(
        Gross_Earned_Premium > 0,
        Paid_Claims_Gross / Gross_Earned_Premium,
        0
      )
    )

  # Add year age (assuming current year is 2024)
  df <- df %>%
    mutate(
      Year_Age = 2024 - Year_of_Account
    )

  # Maturity indicators
  df <- df %>%
    mutate(
      Is_Mature = Year_Age >= 3,
      Maturity_Level = cut(
        Year_Age,
        breaks = c(-Inf, 1, 3, 5, Inf),
        labels = c('Immature', 'Developing', 'Mature', 'Very Mature'),
        right = TRUE
      )
    )

  return(df)
}


#' Generate IBNR summary by Year of Account
#'
#' @param data_source Path to the data CSV file
#' @return data.frame with summary by Year of Account
#' @export
get_ibnr_summary_by_yoa <- function(data_source = '../../synthetic_data/rra_291_gross_premium_ibnr.csv') {

  df <- process_rra_291(data_source)

  summary <- df %>%
    group_by(Year_of_Account) %>%
    summarise(
      Gross_Written_Premium = sum(Gross_Written_Premium, na.rm = TRUE),
      Gross_Earned_Premium = sum(Gross_Earned_Premium, na.rm = TRUE),
      Paid_Claims_Gross = sum(Paid_Claims_Gross, na.rm = TRUE),
      Case_Reserves_Gross = sum(Case_Reserves_Gross, na.rm = TRUE),
      IBNR_Best_Estimate = sum(IBNR_Best_Estimate, na.rm = TRUE),
      IBNR_High = sum(IBNR_High, na.rm = TRUE),
      IBNR_Low = sum(IBNR_Low, na.rm = TRUE),
      Total_Incurred_Gross = sum(Total_Incurred_Gross, na.rm = TRUE),
      Ultimate_Loss_Ratio = mean(Ultimate_Loss_Ratio, na.rm = TRUE),
      .groups = 'drop'
    )

  # Calculate aggregated metrics
  summary <- summary %>%
    mutate(
      Loss_Ratio = ifelse(
        Gross_Earned_Premium > 0,
        Total_Incurred_Gross / Gross_Earned_Premium,
        0
      ),
      IBNR_to_Premium_Ratio = ifelse(
        Gross_Earned_Premium > 0,
        IBNR_Best_Estimate / Gross_Earned_Premium,
        0
      )
    )

  # Add year age
  summary <- summary %>%
    mutate(
      Year_Age = 2024 - Year_of_Account
    )

  # Convert to millions
  summary <- summary %>%
    mutate(
      Gross_Written_Premium_M = Gross_Written_Premium / 1000000,
      Gross_Earned_Premium_M = Gross_Earned_Premium / 1000000,
      Paid_Claims_Gross_M = Paid_Claims_Gross / 1000000,
      Case_Reserves_Gross_M = Case_Reserves_Gross / 1000000,
      IBNR_Best_Estimate_M = IBNR_Best_Estimate / 1000000,
      IBNR_High_M = IBNR_High / 1000000,
      IBNR_Low_M = IBNR_Low / 1000000,
      Total_Incurred_Gross_M = Total_Incurred_Gross / 1000000
    ) %>%
    select(-c(Gross_Written_Premium, Gross_Earned_Premium, Paid_Claims_Gross,
              Case_Reserves_Gross, IBNR_Best_Estimate, IBNR_High, IBNR_Low,
              Total_Incurred_Gross))

  # Round to 2 decimal places
  summary <- summary %>%
    mutate(across(where(is.numeric), ~round(.x, 2)))

  return(summary)
}


#' Generate IBNR summary by Line of Business
#'
#' @param data_source Path to the data CSV file
#' @return data.frame with summary by LOB
#' @export
get_ibnr_summary_by_lob <- function(data_source = '../../synthetic_data/rra_291_gross_premium_ibnr.csv') {

  df <- process_rra_291(data_source)

  summary <- df %>%
    group_by(LOB_Code) %>%
    summarise(
      Gross_Written_Premium = sum(Gross_Written_Premium, na.rm = TRUE),
      Gross_Earned_Premium = sum(Gross_Earned_Premium, na.rm = TRUE),
      Paid_Claims_Gross = sum(Paid_Claims_Gross, na.rm = TRUE),
      Case_Reserves_Gross = sum(Case_Reserves_Gross, na.rm = TRUE),
      IBNR_Best_Estimate = sum(IBNR_Best_Estimate, na.rm = TRUE),
      IBNR_High = sum(IBNR_High, na.rm = TRUE),
      IBNR_Low = sum(IBNR_Low, na.rm = TRUE),
      Total_Incurred_Gross = sum(Total_Incurred_Gross, na.rm = TRUE),
      Ultimate_Loss_Ratio = mean(Ultimate_Loss_Ratio, na.rm = TRUE),
      IBNR_Ratio = mean(IBNR_Ratio, na.rm = TRUE),
      .groups = 'drop'
    )

  # Calculate aggregated metrics
  summary <- summary %>%
    mutate(
      Loss_Ratio = ifelse(
        Gross_Earned_Premium > 0,
        Total_Incurred_Gross / Gross_Earned_Premium,
        0
      )
    )

  # Convert to millions
  summary <- summary %>%
    mutate(
      Gross_Written_Premium_M = Gross_Written_Premium / 1000000,
      Gross_Earned_Premium_M = Gross_Earned_Premium / 1000000,
      Paid_Claims_Gross_M = Paid_Claims_Gross / 1000000,
      Case_Reserves_Gross_M = Case_Reserves_Gross / 1000000,
      IBNR_Best_Estimate_M = IBNR_Best_Estimate / 1000000,
      IBNR_High_M = IBNR_High / 1000000,
      IBNR_Low_M = IBNR_Low / 1000000,
      Total_Incurred_Gross_M = Total_Incurred_Gross / 1000000
    ) %>%
    select(-c(Gross_Written_Premium, Gross_Earned_Premium, Paid_Claims_Gross,
              Case_Reserves_Gross, IBNR_Best_Estimate, IBNR_High, IBNR_Low,
              Total_Incurred_Gross))

  # Round to 2 decimal places
  summary <- summary %>%
    mutate(across(where(is.numeric), ~round(.x, 2)))

  return(summary)
}


#' Analyze IBNR estimate ranges
#'
#' @param data_source Path to the data CSV file
#' @return data.frame with IBNR range analysis
#' @export
get_ibnr_range_analysis <- function(data_source = '../../synthetic_data/rra_291_gross_premium_ibnr.csv') {

  df <- process_rra_291(data_source)

  analysis <- df %>%
    group_by(Year_of_Account, LOB_Code) %>%
    summarise(
      IBNR_Best_Estimate = sum(IBNR_Best_Estimate, na.rm = TRUE),
      IBNR_High = sum(IBNR_High, na.rm = TRUE),
      IBNR_Low = sum(IBNR_Low, na.rm = TRUE),
      IBNR_Range = sum(IBNR_Range, na.rm = TRUE),
      IBNR_Range_Pct = mean(IBNR_Range_Pct, na.rm = TRUE),
      .groups = 'drop'
    )

  # Calculate coefficient of variation
  analysis <- analysis %>%
    mutate(
      IBNR_CV = ifelse(
        IBNR_Best_Estimate > 0,
        IBNR_Range / (2 * IBNR_Best_Estimate),
        0
      )
    )

  # Categorize uncertainty
  analysis <- analysis %>%
    mutate(
      Uncertainty_Level = cut(
        IBNR_Range_Pct,
        breaks = c(0, 0.2, 0.4, 0.6, Inf),
        labels = c('Low', 'Moderate', 'High', 'Very High'),
        right = FALSE
      )
    )

  # Convert to millions
  analysis <- analysis %>%
    mutate(
      IBNR_Best_Estimate_M = IBNR_Best_Estimate / 1000000,
      IBNR_High_M = IBNR_High / 1000000,
      IBNR_Low_M = IBNR_Low / 1000000,
      IBNR_Range_M = IBNR_Range / 1000000
    ) %>%
    select(-c(IBNR_Best_Estimate, IBNR_High, IBNR_Low, IBNR_Range))

  # Round to 2 decimal places
  analysis <- analysis %>%
    mutate(across(where(is.numeric), ~round(.x, 2)))

  return(analysis)
}


#' Analyze ultimate loss ratio trends
#'
#' @param data_source Path to the data CSV file
#' @return data.frame with loss ratio trends by year and LOB
#' @export
get_ultimate_loss_ratio_trend <- function(data_source = '../../synthetic_data/rra_291_gross_premium_ibnr.csv') {

  df <- process_rra_291(data_source)

  trend <- df %>%
    group_by(Year_of_Account, LOB_Code) %>%
    summarise(
      Gross_Earned_Premium = sum(Gross_Earned_Premium, na.rm = TRUE),
      Total_Incurred_Gross = sum(Total_Incurred_Gross, na.rm = TRUE),
      Ultimate_Loss_Ratio = mean(Ultimate_Loss_Ratio, na.rm = TRUE),
      Loss_Ratio_Incurred = mean(Loss_Ratio_Incurred, na.rm = TRUE),
      Year_Age = first(Year_Age),
      .groups = 'drop'
    )

  # Calculate actual loss ratio from totals
  trend <- trend %>%
    mutate(
      Actual_Loss_Ratio = ifelse(
        Gross_Earned_Premium > 0,
        Total_Incurred_Gross / Gross_Earned_Premium,
        0
      )
    )

  # Identify outliers (loss ratio > 1.0)
  trend <- trend %>%
    mutate(
      Is_Outlier = Actual_Loss_Ratio > 1.0
    )

  # Round to 4 decimal places
  trend <- trend %>%
    mutate(across(where(is.numeric), ~round(.x, 4)))

  return(trend)
}


# Main execution for testing
if (sys.nframe() == 0) {
  cat("Processing RRA 291 Gross Premium and IBNR Data...\n")
  df <- process_rra_291()
  cat(sprintf("\nProcessed %d records\n", nrow(df)))

  cat("\n", rep("=", 80), "\n", sep = "")
  cat("IBNR Summary by Year of Account:\n")
  cat(rep("=", 80), "\n", sep = "")
  print(get_ibnr_summary_by_yoa())

  cat("\n", rep("=", 80), "\n", sep = "")
  cat("IBNR Summary by Line of Business:\n")
  cat(rep("=", 80), "\n", sep = "")
  print(get_ibnr_summary_by_lob())

  cat("\n", rep("=", 80), "\n", sep = "")
  cat("IBNR Range Analysis (Sample):\n")
  cat(rep("=", 80), "\n", sep = "")
  print(head(get_ibnr_range_analysis(), 10))
}

File: python_scripts/forms/rra_291_gross_premium_ibnr.py
"""
RRA 291 Gross Premium and IBNR Form - Power BI Python Script
This script processes RRA 291 Gross Premium and IBNR estimates
"""

import pandas as pd
import numpy as np


def process_rra_291(data_source='../../synthetic_data/rra_291_gross_premium_ibnr.csv'):
    """
    Process RRA 291 Gross Premium and IBNR data for Power BI

    Parameters:
    -----------
    data_source : str
        Path to the data CSV file

    Returns:
    --------
    pandas.DataFrame
        Processed data ready for Power BI visualization
    """

    # Load data
    df = pd.read_csv(data_source)

    # Calculate total incurred
    df['Total_Incurred_Gross'] = (
        df['Paid_Claims_Gross'] +
        df['Case_Reserves_Gross'] +
        df['IBNR_Best_Estimate']
    )

    # Calculate reserve ratios
    df['Case_Reserve_Ratio'] = np.where(
        df['Total_Incurred_Gross'] > 0,
        df['Case_Reserves_Gross'] / df['Total_Incurred_Gross'],
        0
    )

    df['IBNR_Ratio'] = np.where(
        df['Total_Incurred_Gross'] > 0,
        df['IBNR_Best_Estimate'] / df['Total_Incurred_Gross'],
        0
    )

    df['Paid_Ratio'] = np.where(
        df['Total_Incurred_Gross'] > 0,
        df['Paid_Claims_Gross'] / df['Total_Incurred_Gross'],
        0
    )

    # Calculate IBNR range
    df['IBNR_Range'] = df['IBNR_High'] - df['IBNR_Low']
    df['IBNR_Range_Pct'] = np.where(
        df['IBNR_Best_Estimate'] > 0,
        df['IBNR_Range'] / df['IBNR_Best_Estimate'],
        0
    )

    # Calculate variance from best estimate
    df['IBNR_High_Variance'] = df['IBNR_High'] - df['IBNR_Best_Estimate']
    df['IBNR_Low_Variance'] = df['IBNR_Best_Estimate'] - df['IBNR_Low']

    # Calculate ultimate loss and loss ratio
    df['Ultimate_Loss'] = df['Total_Incurred_Gross']
    df['Earned_Premium_Ratio'] = np.where(
        df['Gross_Written_Premium'] > 0,
        df['Gross_Earned_Premium'] / df['Gross_Written_Premium'],
        0
    )

    # Calculate loss ratio metrics
    df['Loss_Ratio_Incurred'] = np.where(
        df['Gross_Earned_Premium'] > 0,
        df['Total_Incurred_Gross'] / df['Gross_Earned_Premium'],
        0
    )

    df['Loss_Ratio_Paid'] = np.where(
        df['Gross_Earned_Premium'] > 0,
        df['Paid_Claims_Gross'] / df['Gross_Earned_Premium'],
        0
    )

    # Add year age (assuming current year is 2024)
    df['Year_Age'] = 2024 - df['Year_of_Account']

    # Maturity indicators
    df['Is_Mature'] = df['Year_Age'] >= 3
    df['Maturity_Level'] = pd.cut(
        df['Year_Age'],
        bins=[-1, 1, 3, 5, 100],
        labels=['Immature', 'Developing', 'Mature', 'Very Mature']
    )

    return df


def get_ibnr_summary_by_yoa(data_source='../../synthetic_data/rra_291_gross_premium_ibnr.csv'):
    """
    Generate IBNR summary by Year of Account

    Returns:
    --------
    pandas.DataFrame
        Summary by Year of Account
    """

    df = process_rra_291(data_source)

    summary = df.groupby('Year_of_Account').agg({
        'Gross_Written_Premium': 'sum',
        'Gross_Earned_Premium': 'sum',
        'Paid_Claims_Gross': 'sum',
        'Case_Reserves_Gross': 'sum',
        'IBNR_Best_Estimate': 'sum',
        'IBNR_High': 'sum',
        'IBNR_Low': 'sum',
        'Total_Incurred_Gross': 'sum',
        'Ultimate_Loss_Ratio': 'mean'
    }).reset_index()

    # Calculate aggregated metrics
    summary['Loss_Ratio'] = np.where(
        summary['Gross_Earned_Premium'] > 0,
        summary['Total_Incurred_Gross'] / summary['Gross_Earned_Premium'],
        0
    )

    summary['IBNR_to_Premium_Ratio'] = np.where(
        summary['Gross_Earned_Premium'] > 0,
        summary['IBNR_Best_Estimate'] / summary['Gross_Earned_Premium'],
        0
    )

    # Add year age
    summary['Year_Age'] = 2024 - summary['Year_of_Account']

    # Convert to millions
    for col in ['Gross_Written_Premium', 'Gross_Earned_Premium', 'Paid_Claims_Gross',
                'Case_Reserves_Gross', 'IBNR_Best_Estimate', 'IBNR_High', 'IBNR_Low',
                'Total_Incurred_Gross']:
        summary[f'{col}_M'] = summary[col] / 1000000
        summary = summary.drop(columns=[col])

    summary = summary.round(2)

    return summary


def get_ibnr_summary_by_lob(data_source='../../synthetic_data/rra_291_gross_premium_ibnr.csv'):
    """
    Generate IBNR summary by Line of Business

    Returns:
    --------
    pandas.DataFrame
        Summary by LOB
    """

    df = process_rra_291(data_source)

    summary = df.groupby('LOB_Code').agg({
        'Gross_Written_Premium': 'sum',
        'Gross_Earned_Premium': 'sum',
        'Paid_Claims_Gross': 'sum',
        'Case_Reserves_Gross': 'sum',
        'IBNR_Best_Estimate': 'sum',
        'IBNR_High': 'sum',
        'IBNR_Low': 'sum',
        'Total_Incurred_Gross': 'sum',
        'Ultimate_Loss_Ratio': 'mean',
        'IBNR_Ratio': 'mean'
    }).reset_index()

    # Calculate aggregated metrics
    summary['Loss_Ratio'] = np.where(
        summary['Gross_Earned_Premium'] > 0,
        summary['Total_Incurred_Gross'] / summary['Gross_Earned_Premium'],
        0
    )

    # Convert to millions
    for col in ['Gross_Written_Premium', 'Gross_Earned_Premium', 'Paid_Claims_Gross',
                'Case_Reserves_Gross', 'IBNR_Best_Estimate', 'IBNR_High', 'IBNR_Low',
                'Total_Incurred_Gross']:
        summary[f'{col}_M'] = summary[col] / 1000000
        summary = summary.drop(columns=[col])

    summary = summary.round(2)

    return summary


def get_ibnr_range_analysis(data_source='../../synthetic_data/rra_291_gross_premium_ibnr.csv'):
    """
    Analyze IBNR estimate ranges

    Returns:
    --------
    pandas.DataFrame
        IBNR range analysis
    """

    df = process_rra_291(data_source)

    analysis = df.groupby(['Year_of_Account', 'LOB_Code']).agg({
        'IBNR_Best_Estimate': 'sum',
        'IBNR_High': 'sum',
        'IBNR_Low': 'sum',
        'IBNR_Range': 'sum',
        'IBNR_Range_Pct': 'mean'
    }).reset_index()

    # Calculate coefficient of variation
    analysis['IBNR_CV'] = np.where(
        analysis['IBNR_Best_Estimate'] > 0,
        analysis['IBNR_Range'] / (2 * analysis['IBNR_Best_Estimate']),
        0
    )

    # Categorize uncertainty
    analysis['Uncertainty_Level'] = pd.cut(
        analysis['IBNR_Range_Pct'],
        bins=[0, 0.2, 0.4, 0.6, 100],
        labels=['Low', 'Moderate', 'High', 'Very High']
    )

    # Convert to millions
    for col in ['IBNR_Best_Estimate', 'IBNR_High', 'IBNR_Low', 'IBNR_Range']:
        analysis[f'{col}_M'] = analysis[col] / 1000000
        analysis = analysis.drop(columns=[col])

    analysis = analysis.round(2)

    return analysis


def get_ultimate_loss_ratio_trend(data_source='../../synthetic_data/rra_291_gross_premium_ibnr.csv'):
    """
    Analyze ultimate loss ratio trends

    Returns:
    --------
    pandas.DataFrame
        Loss ratio trends by year and LOB
    """

    df = process_rra_291(data_source)

    trend = df.groupby(['Year_of_Account', 'LOB_Code']).agg({
        'Gross_Earned_Premium': 'sum',
        'Total_Incurred_Gross': 'sum',
        'Ultimate_Loss_Ratio': 'mean',
        'Loss_Ratio_Incurred': 'mean',
        'Year_Age': 'first'
    }).reset_index()

    # Calculate actual loss ratio from totals
    trend['Actual_Loss_Ratio'] = np.where(
        trend['Gross_Earned_Premium'] > 0,
        trend['Total_Incurred_Gross'] / trend['Gross_Earned_Premium'],
        0
    )

    # Identify outliers (loss ratio > 1.0)
    trend['Is_Outlier'] = trend['Actual_Loss_Ratio'] > 1.0

    trend = trend.round(4)

    return trend


# For use in Power BI
# df = process_rra_291()


if __name__ == "__main__":
    # Test the functions
    print("Processing RRA 291 Gross Premium and IBNR Data...")
    df = process_rra_291()
    print(f"\nProcessed {len(df)} records")

    print("\n" + "="*80)
    print("IBNR Summary by Year of Account:")
    print("="*80)
    print(get_ibnr_summary_by_yoa())

    print("\n" + "="*80)
    print("IBNR Summary by Line of Business:")
    print("="*80)
    print(get_ibnr_summary_by_lob())

    print("\n" + "="*80)
    print("IBNR Range Analysis (Sample):")
    print("="*80)
    print(get_ibnr_range_analysis().head(10))
